Mathlib/Algebra/MonoidAlgebra/Degree.lean

```diff
@@ -477,25 +486,48 @@ lemma Monic.supDegree_mul_of_ne_zero
     have := covariantClass_le_of_lt B B (Function.swap (· + ·))
     exact fun a ha => (Finset.le_sup ha).trans (supDegree_mul_le hadd)
 
+lemma Monic.supDegree_mul_of_ne_zero_right
+    (hD : D.Injective) (hadd : ∀ a1 a2, D (a1 + a2) = D a1 + D a2)
+    (hp : p.Monic D) (hq : q ≠ 0) :
+    (p * q).supDegree D = p.supDegree D + q.supDegree D := by
+  cases subsingleton_or_nontrivial R; · exact (hq (Subsingleton.elim _ _)).elim
+  apply supDegree_eq_of_max
+  · rw [← AddSubsemigroup.coe_set_mk (Set.range D), ← AddHom.srange_mk _ hadd, SetLike.mem_coe]
+    exact add_mem (supDegree_mem_range D hp.ne_zero) (supDegree_mem_range D hq)
+  · simp_rw [Finsupp.mem_support_iff, apply_supDegree_add_supDegree hD hadd, hp, one_mul,
+      Ne, leadingCoeff_eq_zero hD, hq, not_false_eq_true]
+  · have := covariantClass_le_of_lt B B (· + ·)
+    have := covariantClass_le_of_lt B B (Function.swap (· + ·))
```

Comment:
Is this really needed?

Reply:
Seems to be, in the sense that removing those lines complains. I've been trying very hard to avoid learning about these classes, so can't say anything more.